src/backend/catalog/catalog.c

```diff
@@ -558,10 +665,31 @@ GetNewRelFileNode(Oid reltablespace, Relation pg_class, char relpersistence)
 
 		/* Generate the OID */
 		if (pg_class)
-			rnode.node.relNode = GetNewOidWithIndex(pg_class, ClassOidIndexId,
+		{
+			if (relpersistence == RELPERSISTENCE_TEMP && sql_dialect == SQL_DIALECT_TSQL)
+			{
+				/* Temp tables use temp OID logic */
+				rnode.node.relNode = GetNewTempOidWithIndex(pg_class, ClassOidIndexId,
```

Comment:
why have we written this new method? wouldn't GetNewOidWithIndex handle both cases?

Reply:
GetNewTempOidWithIndex calls GetNewTempObjectId instead of GetNewObjectId. We could merge the two, but it would require a refactoring of GetNewTempOidWithIndex to know relpersistence of the relation, right? This way is less invasive to engine code IMO